x-pack/plugins/reporting/public/panel_actions/get_csv_panel_action.tsx

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import dateMath from '@elastic/datemath';
+import { i18n } from '@kbn/i18n';
+
+import { ContextMenuAction, ContextMenuActionsRegistryProvider } from 'ui/embeddable';
+import { PanelActionAPI } from 'ui/embeddable/context_menu_actions/types';
+import { kfetch } from 'ui/kfetch';
+import { toastNotifications } from 'ui/notify';
+
+const API_BASE_URL = '/api/reporting/v1/generate/immediate/csv/saved-object/';
+
+class GetCsvReportPanelAction extends ContextMenuAction {
+  constructor() {
+    super(
+      {
+        displayName: i18n.translate('xpack.reporting.dashboard.downloadCsvPanelTitle', {
+          defaultMessage: 'Download CSV',
+        }),
+        id: 'downloadCsvReport',
+        parentPanelId: 'mainMenu',
+      },
+      {
+        icon: 'document',
+      }
+    );
+  }
+
+  public async generateJobParams({ searchEmbeddable }: { searchEmbeddable: any }) {
+    const adapters = searchEmbeddable.getInspectorAdapters();
+    if (!adapters) {
+      return {};
+    }
+
+    if (adapters.requests.requests.length === 0) {
+      return {};
+    }
+
+    return searchEmbeddable.searchScope.searchSource.getSearchRequestBody();
+  }
+
+  public isVisible = (panelActionAPI: PanelActionAPI): boolean => {
+    const { embeddable } = panelActionAPI;
+
+    if (embeddable && embeddable.hasOwnProperty('savedSearch')) {
+      return true;
+    }
+
+    return false;
+  };
+
+  public onClick = async (panelActionAPI: PanelActionAPI) => {
+    const { embeddable } = panelActionAPI as any;
+    const {
+      timeRange: { from, to },
+    } = embeddable;
+
+    if (!embeddable) {
+      return;
+    }
+
+    const searchEmbeddable = embeddable;
+    const state = await this.generateJobParams({ searchEmbeddable });
+
+    const id = `search:${embeddable.savedSearch.id}`;
+    const filename = embeddable.savedSearch.title;
```

Comment:
Maybe next, we should work on exposing the SearchEmbeddable interface here, because `.savedSearch` and `.$rootScope` are typed as private in `src/legacy/core_plugins/kibana/public/discover/embeddable/search_embeddable`

Reply:
In addition to the other headaches with the `SearchEmbeddable` interface (😁) it seems too limited for the definition to be in `/discover`. The Discover app is used for creating a saved search... but having the defined structure of a saved search is important all over the Kibana codebase.

Reply:
Yeah, I think I'm going to add another issue where we do some cleanup/migration of this component so we can better use it in our x-pack code.